### PR TITLE
Tree Manager and Display UI. Code updated for modularity

### PR DESCRIPTION
**Pull request by mayank33**

### CODE REVIEW

File: src/Display.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#include <Wt/WGridLayout>
+#include <Wt/WGroupBox>
+#include <Wt/WVBoxLayout>
+#include <Wt/WHBoxLayout>
+#include <Wt/WImage>
+#include <Wt/WCheckBox>
+#include <Wt/WLineEdit>
+#include <Wt/WText>
+#include <Wt/WLabel>
+#include <Wt/WSelectionBox>
```
Comment (rashadkm) [inline src/Display.h:13]: only include required headers.

File: src/LayerManager.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include <Wt/WGroupBox>
+#include <Wt/WVBoxLayout>
+#include <Wt/WHBoxLayout>
+#include <Wt/WImage>
+#include <Wt/WCheckBox>
+#include <Wt/WLineEdit>
+#include <Wt/WText>
+#include <Wt/WLabel>
+#include <Wt/WSelectionBox>
+
```
Comment (rashadkm) [inline src/LayerManager.h:19]: same as above. check if all these includes are required. 

put includes in .h and include the class's header inisde the .cpp

in startup.cpp you can include startup.h and all those wt based includes go into startup.h.

remember some of them are required to included in .cpp file also. So check them

File: src/Toolbar.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#include <Wt/WGroupBox>
+#include <Wt/WVBoxLayout>
+#include <Wt/WHBoxLayout>
+#include <Wt/WImage>
+#include <Wt/WCheckBox>
+#include <Wt/WLineEdit>
+#include <Wt/WText>
+#include <Wt/WLabel>
+#include <Wt/WSelectionBox>
+
```
Comment (rashadkm) [inline src/Toolbar.h:14]: same as last two comments

**Comment by rashadkm** [comment]
fix the issues in the comments and i will merge it.

**Comment by mayank33** [comment]
updated. some headers files are not removed as they may be used in future

File: src/MainUI.cpp [hunk 4]
```diff
@@ -54,15 +59,39 @@ void MainUI::createUI(Wt::WContainerWidget *parent) {
...
           pugi::xml_node menu_items_menu_items_node = menu_items_menu_node.child("items");
           while ( menu_items_menu_items_node ) {
              /* iteration over internal occurance of menuitem */
              pugi::xml_node menu_items_menu_items_menuitem_node = menu_items_menu_items_node.child("menuitem");
              while( menu_items_menu_items_menuitem_node ) {
-                nextLevel->addItem(menu_items_menu_items_menuitem_node.child_value("label"));
+                //string* tt="haha";
+                //nextLevel->addItem(menu_items_menu_items_menuitem_node.child_value("label"));
+                //nextLevel::m_grassModuleName = menu_items_menu_items_menuitem_node.child_value("command");
+              void *vague_pointer;
```
Comment (rashadkm) [inline src/MainUI.cpp:69]: can you explain this part?

File: src/MainUI.cpp [hunk 5]
```diff
@@ -127,3 +156,43 @@ void MainUI::createUI(Wt::WContainerWidget *parent) {
...
+   string s="";
+   s.clear();
+   while(*(item1+i)!='\0')
+              {
+                s.push_back(*(item1+i));
+                i++;
+              }
+ cout<<s<<endl;
+cout<<dd->data();
+if(dd->text()=="Digitize vector map using Tcl/Tk digitizer")
```
Comment (rashadkm) [inline src/MainUI.cpp]: this is ugly. you are saying to do an if check for all grass menu items ?

File: src/MenuItem.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+#include <Wt/WContainerWidget>
+#include <Wt/WHBoxLayout>
+#include <Wt/WVBoxLayout>
+#include <Wt/WPushButton>
+#include <Wt/WCheckBox>
+#include <Wt/WSelectionBox>
+#include <Wt/WApplication>
+#include <Wt/WBreak>
+#include <Wt/WMenuItem>
+#include <Wt/WPopupMenu>
```
Comment (rashadkm) [inline src/MenuItem.h]: again, you are doing the same. copy headers include from existing file. Please please do include only necessary. It had warned before

File: src/MainUI.cpp [hunk 7]
```diff
@@ -54,15 +59,39 @@ void MainUI::createUI(Wt::WContainerWidget *parent) {
        pugi::xml_node menu_items_menu_node = menu_items_node.child("menu");
        while( menu_items_menu_node ) {
           /* submenu creation */
-          Wt::WPopupMenu *nextLevel = new Wt::WPopupMenu();
+          MenuItem *nextLevel = new MenuItem();
```
Comment (rashadkm) [inline src/MainUI.cpp]: this is not the menuitem class

File: src/MainUI.cpp [hunk 8]
```diff
@@ -54,15 +59,39 @@ void MainUI::createUI(Wt::WContainerWidget *parent) {
...
           firstLevel->addMenu(menu_items_menu_node.child_value("label"), nextLevel);
           /* iteration over second occurance of items */
           pugi::xml_node menu_items_menu_items_node = menu_items_menu_node.child("items");
           while ( menu_items_menu_items_node ) {
              /* iteration over internal occurance of menuitem */
              pugi::xml_node menu_items_menu_items_menuitem_node = menu_items_menu_items_node.child("menuitem");
              while( menu_items_menu_items_menuitem_node ) {
-                nextLevel->addItem(menu_items_menu_items_menuitem_node.child_value("label"));
```
Comment (rashadkm) [inline src/MainUI.cpp:62]: here it must be nextlevel->addItem( menuitem instance)

Comment (rashadkm) [inline src/MainUI.cpp:62]: MenuItem *item = new MenuItem (  name of grass command) 
this name of grass command will be passed to m_grassModuleName.
you can get them later in them->triggered method

File: src/MainUI.cpp [hunk 9]
```diff
@@ -80,145 +112,87 @@ void MainUI::createUI(Wt::WContainerWidget *parent) {
...
-  toolBar->addButton(modeler);
-  toolBar->addButton(composer);
-  toolBar->addButton(scriptload);
-  toolBar->addButton(guisettings);
-  toolBar->addButton(manual);
-
-  addWidget(toolBar);
+
+  Wt::WContainerWidget *toolbarcontainer = new Wt::WContainerWidget();
+  Toolbar* toolbar = new Toolbar(toolbarcontainer);
```
Comment (rashadkm) [inline src/MainUI.cpp:106]: good! . Toolbar is moved to a separate class

**Pull request merged**